### PR TITLE
#29664 one version to rule them all

### PR DESCRIPTION
**Pull request by danchr**
This PR introduces `Location` and `Version` — the supermodels for versions!

### CODE REVIEW

File: doc/requirements-dev.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 -r requirements.txt
 ipython
 pycodestyle==2.4.0
+django_extensions
```
Comment (magenta-af) [inline doc/requirements-dev.txt]: What are these, and why are they useful? :-) I saw no mention of this, and the commit message doesn't say anything either...

File: src/os2datascanner/engine/scanners/scanner_types/scanner.py [hunk 2]
```diff
@@ -18,6 +18,8 @@
 
 import structlog
 
+from os2datascanner.projects.admin.adminapp.models.location_model import \
```
Comment (magenta-af) [inline src/os2datascanner/engine/scanners/scanner_types/scanner.py]: A _backslash line continuation_, you say? 🤨

File: src/os2datascanner/projects/admin/adminapp/migrations/0001_initial.py [hunk 3]
```diff
@@ -1,12 +1,14 @@
 # -*- coding: utf-8 -*-
```
Comment (magenta-af) [inline src/os2datascanner/projects/admin/adminapp/migrations/0001_initial.py]: I can see a case for ignoring backwards compatibility and for squashing all the migrations into a new `0001_initial` before the release, but do we really need to do this repeatedly in every branch? (And won't this clash with #56?)

File: src/os2datascanner/projects/admin/adminapp/models/version_model.py [hunk 4]
```diff
@@ -15,25 +15,35 @@
...
 class Version(models.Model):
 
+    objects = InheritanceManager()
+
     """A representation of an actual URL on a domain with its MIME type."""
 
-    url = models.CharField(max_length=2048, verbose_name='URL')
-    scan = models.ForeignKey(Scan, null=False, verbose_name='Scan',
-                             related_name='urls',
+    location = models.ForeignKey('Location', null=False,
```
Comment (magenta-af) [inline src/os2datascanner/projects/admin/adminapp/models/version_model.py]: (This is not at all an actionable comment, but this is so _neat_! Being able to easily say "here are all the known versions of this object" is exactly what we need for a lot of the report layouts we discussed. :+1:)

File: src/os2datascanner/projects/admin/adminapp/views/report_views.py [hunk 5]
```diff
@@ -147,37 +147,6 @@ def get_context_data(self, **kwargs):
         except ObjectDoesNotExist:
             pass
 
-        if hasattr(this_scan.scanner, 'filescanner'):
```
Comment (magenta-af) [inline src/os2datascanner/projects/admin/adminapp/views/report_views.py]: I agree (enthusiastically!) that this code should go, but how do reports behave at the moment without it -- do you get raw `file://` URLs?

File: src/os2datascanner/projects/admin/adminapp/migrations/0001_initial.py (hunk 3 above)
Comment (danchr) [reply src/os2datascanner/projects/admin/adminapp/migrations/0001_initial.py]: Well, this one would _also_ be ridiculously hard to write a migration for, so I did another reset. The thinking is that we reset whenever writing the migration is harder than handling the conflict 🙂

File: src/os2datascanner/engine/scanners/scanner_types/scanner.py (hunk 2 above)
Comment (danchr) [reply src/os2datascanner/engine/scanners/scanner_types/scanner.py]: > A _backslash line continuation_, you say? 🤨

Very much not intended; I'm using an IDE now, and it intends to inject those 😞

File: doc/requirements-dev.txt (hunk 1 above)
Comment (danchr) [reply doc/requirements-dev.txt]: [`django-extensions`](https://django-extensions.readthedocs.io) adds all sorts of useful commands for development, such as `reset_db` and `graph_models`.

File: src/os2datascanner/projects/admin/adminapp/views/report_views.py (hunk 5 above)
Comment (danchr) [reply src/os2datascanner/projects/admin/adminapp/views/report_views.py]: I'll check…

Comment (danchr) [reply src/os2datascanner/projects/admin/adminapp/views/report_views.py]: …and it looks like this, now:

image

Comment (danchr) [reply src/os2datascanner/projects/admin/adminapp/views/report_views.py]: > …and it looks like this, now:
> 
> [snip]

Much better:

image

Comment (magenta-af) [reply src/os2datascanner/projects/admin/adminapp/views/report_views.py]: :+1:

**Pull request merged**